Hijack triggers

GANDAMU opened the pull request:
hi
team
Hey team.
TEAM.
HEY TEAM.

I think...this might actually give us the piece we need for interactivity? Hopefully this is the branch that gets us to meaningful progress on the Pong game?

The "test" sequence:

1. `oscsend localhost 5006 /trigger/left_nail` - See that left_nail triggers "braindead"
2. `oscsend localhost 5006 /rosary/add_effect/name/trigger_hijack_demo` - When "braindead" ends, do this and see the demo effect
3. `oscsend localhost 5006 /trigger/left_nail` - See that the left_nail now reverses the "looper" effect's direction when fired when the red looper is on the left-side whitened bead
4. `oscsend localhost 5006 /rosary/clear_effects` - Remove the demo effect
5. `oscsend localhost 5006 /trigger/left_nail` - See that left_nail once again triggers "braindead"

GANDAMU committed:
- Just so I can link a line

GANDAMU committed:
- Some cleanup

GANDAMU committed:
- Move trigger_hijack stuff up into effect.py, for repeatability
- Add some cleanup of trigger hijacks in rosary

GANDAMU committed:
- Get old bounce.py (hopefully I did it in a way that merging down master won't see my changes as an update)

GANDAMU committed:
- Well that wasn't as bad as I was afraid it'd be

Merge branch 'master' of github.com:jdblair/MegaPrayer-oscled into hijack_triggers

Conflicts:
	python/mp/effects/__init__.py

jdblair commented on python/mp/rosary.py:
```diff
@@ -535,22 +548,38 @@ def mainloop(self, *args, **kwargs):
+
+        else:
+
+            requested_trigger = self.trigger_registry.get(trigger_name)
```
this is the original behavior, right?

GANDAMU replied:
Yep, I just shoved the entirety of the old method into the else.

jdblair commented:
I like this, I think its awesome. I don't have specific comments yet, but it gave me an idea how to write the next feature we need: "idle" behavior. E.g., what is the rosary doing when its just sitting there? I was struggling with how to design this, but now I see we can just write an effect that handles it, and processes input to decide what to do, and start other effects.

jdblair commented on python/mp/effects/__init__.py:
```diff
@@ -23,5 +23,6 @@
-    'random_fill'
+    'random_fill',
+    'trigger_hijack_demo'
 ]
```
If you add your new effect in the middle of the list then we won't have conflicts when merging multiple new effects.

GANDAMU replied:
If we had more time I'd ask "lordy is there no way to just import all the classes and be done with it?"

jdblair commented on python/mp/rosary.py:
```diff
@@ -535,22 +548,38 @@ def mainloop(self, *args, **kwargs):
+            # if many effects hijack the same trigger, then only the newest
+            # effect's hijacked trigger fires
+            hijacked_effect, hijacked_method = hijacked[-1]
+
```
Does this mean I can do this?
1) Effect A hijacks the triggers, and at some point starts effect B.
2) Effect B hijacks the triggers and does stuff.
3) Effect B exits (sets self.finished=true), Effect A processes triggers again.

GANDAMU replied:
That is the intention.

Although I've only tested 1 "layer" so far (if you consider what you're proposing as having 2 "layers")

jdblair commented:
+1. Merge it. My only objection is "hijacking" is a weird term to use, but I'm getting used to it.

GANDAMU reviewed (commented):
Okay, gonna fix that conflict and merge it in.

GANDAMU committed:
- __init__.py thing

Merge branch 'master' of github.com:jdblair/MegaPrayer-oscled into hijack_triggers

Conflicts:
	python/mp/effects/__init__.py

GANDAMU commented:
Also I love the term "hijacking" here. It feels so COOL AND HACKERY.

![https://media.giphy.com/media/uPnKU86sFa2fm/giphy.gif](https://media.giphy.com/media/uPnKU86sFa2fm/giphy.gif)
